Add the citemgroup command

xpple opened the pull request:
Closes #235 

This. Was. A. Lot. Of. Pain.

This took me absolutely ages to code, but I'm finally done. Dynamically changing item groups is HARD. I initally got this working with `ItemGroup#appendStacks()`, but that turned out to be abuse of the method. After that was no longer an option, I tried setting the `group` for an `Item`, which removed the item from the vanilla groups 🤣. The last option was to recreate the entire item group to "modify" it. While this may sound inefficient, I did manage to optimise it by only recreating the item groups >= the to be modified group.

xpple committed:
Add the citemgroup command

xpple committed:
Merge remote-tracking branch 'origin/fabric' into citemgroup

xpple committed:
Fix styling

xpple committed:
Catch InvalidIdentifierException + substring the item group name

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/command/ItemGroupCommand.java:
```diff
@@ -0,0 +1,307 @@
+        try {
+            loadFile();
+        } catch (IOException e) {
+            LOGGER.info("Could not load groups file, hence /citemgroup will not work!");
```
```suggestion
            LOGGER.error("Could not load groups file, hence /citemgroup will not work!", e);
```

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/command/ItemGroupCommand.java:
```diff
@@ -0,0 +1,307 @@
+        addClientSideCommand("citemgroup");
+
+        LiteralCommandNode<ServerCommandSource> citemgroup = dispatcher.register(literal("citemgroup"));
+        dispatcher.register(literal("citemgroup")
```
Can some of these indents below are twice as much as they should be, code shouldn't the this far right!

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/mixin/CreativeInventoryScreenAccessor.java:
```diff
@@ -0,0 +1,19 @@
+        throw new AssertionError();
+    }
+
+    @SuppressWarnings("AccessorTarget")
```
You can use `@Dynamic` here instead

Earthcomputer replied:
Why was this resolved?

xpple replied:
Because I added the `@Dynamic` annotation. For your information, if I remember correctly, the `@Dynamic` annotation can't be used _instead of_ the `@SuppressWarnings` annotation: it's complementairy.

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/mixin/MixinItemGroup.java:
```diff
@@ -0,0 +1,40 @@
+@Mixin(ItemGroup.class)
+public abstract class MixinItemGroup implements IItemGroup {
+
+    @Shadow @Final @Mutable public static ItemGroup[] GROUPS;
+
+    @Mutable @Shadow @Final private Text translationKey;
```
The order of these annotations is inconsistent

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/mixin/MixinItemGroup.java:
```diff
@@ -0,0 +1,40 @@
+
+    @Inject(method = "<init>", at = @At("TAIL"))
+    private void substring(int index, String id, CallbackInfo ci) {
+        if (id.contains("clientcommands.")) {
```
Maybe `startsWith` instead?

Earthcomputer reviewed (changes requested):
I'm not sure about the whole item-groups-are-lists thing either. I'd prefer if they were sets you can add to, or remove from by an item predicate. Also, subcommands for renaming an item group, changing its icon, and duplicating one would be nice.

xpple committed:
Merge remote-tracking branch 'origin/fabric' into citemgroup

xpple committed:
Shrink indents

xpple commented:
The above commit also implements [this requested change](https://github.com/Earthcomputer/clientcommands/pull/236#discussion_r632073612).

xpple committed:
Put annotations consistent in order

xpple committed:
Use .startsWith rather than .contains

xpple committed:
Add @Dynamic annotation

xpple commented:
Working on this still
> Also, subcommands for renaming an item group, changing its icon, and duplicating one would be nice.

xpple committed:
Implement requested feature

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/command/ItemGroupCommand.java:
```diff
@@ -0,0 +1,361 @@
+
+        groups.remove(name);
+
+        reloadGroups(12);
```
Hardcoding this value 12 seems dangerous.What if other mods add item groups? What if a new Minecraft update adds new item groups? There must be a better way to do this, perhaps by filtering out the clientcommands groups?

xpple replied:
Yeah this is difficult. If other mods would add an item group, it would be deleted. I will try to implement a check for clientcommands-created item groups. I do not worry as much about a Minecraft update adding a new group, that will just change the starting index (12 now).

xpple replied:
Does this look reasonable? Edit: the below code is not functional.
```java
@Override
public void shrink(int index) {
    ItemGroup[] tempGroups = GROUPS;
    tempGroups = Arrays.stream(tempGroups).filter(itemGroup -> !itemGroup.getName().startsWith("clientcommands.")).toArray(ItemGroup[]::new);
    GROUPS = new ItemGroup[index + tempGroups.length];
    System.arraycopy(tempGroups, 0, GROUPS, 0, GROUPS.length);
}
```

xpple replied:
To be honest I'm going quite insane from this mess of arrays and indices.

xpple replied:
Update to clarify: the most recent commit fixed this issue.

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/command/ItemGroupCommand.java:
```diff
@@ -0,0 +1,361 @@
+
+        Group group = groups.get(name);
+        ListTag items = group.getItems();
+        if ((index < 0) || (index > items.size())) {
```
```suggestion
        if ((index < 0) || (index >= items.size())) {
```

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/command/ItemGroupCommand.java:
```diff
@@ -0,0 +1,361 @@
+
+        Group group = groups.get(name);
+        ListTag items = group.getItems();
+        if (index < 0 || index > items.size()) {
```
```suggestion
        if (index < 0 || index >= items.size()) {
```

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/command/ItemGroupCommand.java:
```diff
@@ -0,0 +1,361 @@
+                    .build();
+
+            groups.put(_new, new Group(itemGroup, icon, items));
+        } catch (InvalidIdentifierException e) {
```
Maybe `Identifier.tryParse` could be used here instead of a try-catch?

xpple replied:
That returns `null` if it fails, which is not what we want here, do we?

Earthcomputer replied:
Idk I find it cleaner

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/command/ItemGroupCommand.java:
```diff
@@ -0,0 +1,361 @@
+            File currentFile = new File(configPath.toFile(), "groups.dat");
+            Util.backupAndReplace(currentFile, newFile, backupFile);
+        } catch (IOException e) {
+            throw SAVE_FAILED_EXCEPTION.create();
```
Printing the stack trace here might be useful to help people figure out what's going wrong

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/mixin/MixinItemGroup.java:
```diff
@@ -0,0 +1,40 @@
+
+    @Override
+    public void shrink(int index) {
+        ItemGroup[] tempGroups = GROUPS;
+        GROUPS = new ItemGroup[index];
+        System.arraycopy(tempGroups, 0, GROUPS, 0, GROUPS.length);
```
This can be simplified to `Arrays.copyOf`. Also, was there a reason that `GROUPS` couldn't be accessed by an accessor mixin instead?

xpple replied:
`Arrays.copyOf` calls `System.arraycopy` internally, are you sure you want to use it?

As to the other question, I guess it could be 🤷‍♂️.

xpple replied:
I think I missed your response here, what was it again?

Earthcomputer replied:
The way it's been rewritten, you can just reassign `GROUPS` like I commented below.

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/command/ItemGroupCommand.java:
```diff
@@ -0,0 +1,361 @@
+
+        reloadGroups(itemGroup.getIndex());
+        saveFile();
+        sendFeedback("commands.citemgroup.addStack.success", itemStack.getItem().getName(), name);
```
Is `getName()` the same method that's used to display weapons in death messages?

xpple replied:
Ohh yeah maybe it will return the name of the item if it has been renamed. Let me check that.

xpple replied:
Nope, it won't. This will always return the default name of the item. See the below image for reference:
![image](https://user-images.githubusercontent.com/49305700/119528120-3acee300-bd81-11eb-950e-c23a8fb5f3f7.png)

xpple committed:
Implement requested changes

xpple commented:
I've implemented most of the requested change, only not those who are still to be discussed. Anyway, to solve the issue proposed [here](https://github.com/Earthcomputer/clientcommands/pull/236#discussion_r637886623), I had to get rid of the initial optimisation I implemented. It's either the optimisation or compatibility with other mods that add item groups.

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/mixin/MixinItemGroup.java:
```diff
@@ -0,0 +1,43 @@
+    @Override
+    public void shrink() {
+        ItemGroup[] tempGroups = GROUPS;
+        tempGroups = Arrays.stream(tempGroups).filter(itemGroup -> !itemGroup.getName().startsWith("clientcommands.")).toArray(ItemGroup[]::new);
```
`toArray(Supplier)` is not a Java 8 method

xpple replied:
Yes it is [(see the API documentation)](https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#toArray-java.util.function.IntFunction-).

Earthcomputer replied:
Oh right, I forgot this was a stream rather than a collection

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/mixin/MixinItemGroup.java:
```diff
@@ -0,0 +1,43 @@
+        ItemGroup[] tempGroups = GROUPS;
+        tempGroups = Arrays.stream(tempGroups).filter(itemGroup -> !itemGroup.getName().startsWith("clientcommands.")).toArray(ItemGroup[]::new);
+        GROUPS = new ItemGroup[tempGroups.length];
+        System.arraycopy(tempGroups, 0, GROUPS, 0, GROUPS.length);
```
Why not just assign to `GROUPS` directly rather than using a temp array?

xpple replied:
Like this:
```java
GROUPS = Arrays.stream(GROUPS).filter(itemGroup -> !itemGroup.getName().startsWith("clientcommands.")).toArray(ItemGroup[]::new);
```
?

Earthcomputer replied:
Yep

xpple replied:
Kinda stupid I missed that 🤣, problems with modifying existing code I guess.

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/interfaces/IItemGroup.java:
```diff
@@ -0,0 +1,7 @@
+package net.earthcomputer.clientcommands.interfaces;
+
+public interface IItemGroup {
+    int getLength();
```
`getLength` is never used

xpple committed:
Implement requested changes

Earthcomputer reviewed (approved):
Lgtm
